internal/services/attestor.go

```diff
@@ -0,0 +1,97 @@
+package services
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/rs/zerolog"
+	mt "github.com/txaty/go-merkletree"
+)
+
+var CredentialExpiresAt = "ExprTime"
+var Values = "Values"
+
+type leaf struct {
+	data []byte
+}
+
+func (l *leaf) Serialize() ([]byte, error) {
+	// since OpenZep will sort bytes before hashing
+	// abi encoding and serialization is performed before
+	// we pass any data to the merkle tree func
+	// thus, we can just return the bytes directly here
+	return l.data, nil
+}
+
+func abiEncode(types []string, values []interface{}) ([]byte, error) {
+	if len(types) != len(values) {
+		return nil, fmt.Errorf("type array  and value array must be same length")
+	}
+
+	args := abi.Arguments{}
+	for _, argType := range types {
+		abiType, err := abi.NewType(argType, "", nil)
+		if err != nil {
+			return nil, err
+		}
+		args = append(args, abi.Argument{Type: abiType})
+	}
+
+	return args.Pack(values...)
+}
+
+type Attestor struct {
+	config mt.Config
+	log    *zerolog.Logger
+}
+
+func NewAttestor(logger *zerolog.Logger) *Attestor {
+	return &Attestor{
+		config: mt.Config{
+			SortSiblingPairs: true, // parameter for OpenZeppelin compatibility
+			HashFunc: func(data []byte) ([]byte, error) {
+				return crypto.Keccak256(data), nil
+			},
+			Mode:               mt.ModeProofGenAndTreeBuild,
+			DisableLeafHashing: true, // also required for OpenZeppelin compatilibity
```

Comment:
Both here and on line 22, the point is that we're already serializing and hashing, yeah?

Reply:
yes-- updated the doc string to clarify this. I'm wondering if I should try making a pr to the `"github.com/txaty/go-merkletree"` library